### ISSUE TITLE
Javadoc fails when documenting annotations

### ISSUE THREAD

**Issue by roxspring** [Kotlin/dokka#464]
Given a custom annotation:

```kotlin
package api

annotation class CustomAnnotation(
    val value: String
)
```

And a fairly simple `build.gradle.kts`:

```kotlin
import org.jetbrains.dokka.gradle.DokkaTask

plugins {

    kotlin("jvm") version "1.3.21"

    id("org.jetbrains.dokka") version "0.9.18"
}

repositories {
    jcenter()
}

dependencies {
    compile("org.jetbrains.kotlin:kotlin-stdlib")
}

tasks.withType(DokkaTask::class.java).all {
    outputFormat = "javadoc"
}
```

`gradle dokka` claims success but fails to produce any documentation and clearly suffers a fairly brutal exception:

```
> Task :dokka
No documentation for api.CustomAnnotation (CustomAnnotation.kt:3)
No documentation for api.CustomAnnotation$value (CustomAnnotation.kt:4)
java.lang.ClassCastException: org.jetbrains.dokka.javadoc.ClassDocumentationNodeAdapter cannot be cast to com.sun.javadoc.AnnotationTypeDoc
        at com.sun.tools.doclets.formats.html.HtmlDoclet.generateClassFiles(HtmlDoclet.java:181)
        at com.sun.tools.doclets.internal.toolkit.AbstractDoclet.generateClassFiles(AbstractDoclet.java:194)
        at com.sun.tools.doclets.internal.toolkit.AbstractDoclet.startGeneration(AbstractDoclet.java:139)
        at com.sun.tools.doclets.internal.toolkit.AbstractDoclet.start(AbstractDoclet.java:82)
        at com.sun.tools.doclets.formats.html.HtmlDoclet.start(HtmlDoclet.java:80)
        at org.jetbrains.dokka.javadoc.JavadocGenerator.buildPages(dokka-adapters.kt:16)
        at org.jetbrains.dokka.GeneratorKt.buildAll(Generator.kt:13)
        at org.jetbrains.dokka.GeneratorKt.buildAll(Generator.kt:23)
        at org.jetbrains.dokka.DokkaGenerator.generate(DokkaGenerator.kt:47)
        at org.jetbrains.dokka.DokkaBootstrapImpl.generate(DokkaBootstrapImpl.kt:79)
        at sun.reflect.GeneratedMethodAccessor489.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.jetbrains.dokka.DelegatedInvocationHandler.invoke(ProxyUtils.kt:41)
        at com.sun.proxy.$Proxy61.generate(Unknown Source)
        at org.jetbrains.dokka.gradle.DokkaTask.generate(main.kt:368)
        at sun.reflect.GeneratedMethodAccessor462.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:103)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:48)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:41)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:28)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:701)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:668)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$4.run(ExecuteActionsTaskExecuter.java:337)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:92)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:326)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:311)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.access$200(ExecuteActionsTaskExecuter.java:75)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.execute(ExecuteActionsTaskExecuter.java:158)
        at org.gradle.internal.execution.impl.steps.ExecuteStep.execute(ExecuteStep.java:46)
        at org.gradle.internal.execution.impl.steps.CancelExecutionStep.execute(CancelExecutionStep.java:34)
        at org.gradle.internal.execution.impl.steps.TimeoutStep.executeWithoutTimeout(TimeoutStep.java:69)
        at org.gradle.internal.execution.impl.steps.TimeoutStep.execute(TimeoutStep.java:49)
        at org.gradle.internal.execution.impl.steps.CatchExceptionStep.execute(CatchExceptionStep.java:34)
        at org.gradle.internal.execution.impl.steps.CreateOutputsStep.execute(CreateOutputsStep.java:49)
        at org.gradle.internal.execution.impl.steps.SnapshotOutputStep.execute(SnapshotOutputStep.java:42)
        at org.gradle.internal.execution.impl.steps.SnapshotOutputStep.execute(SnapshotOutputStep.java:28)
        at org.gradle.internal.execution.impl.steps.CacheStep.executeWithoutCache(CacheStep.java:136)
        at org.gradle.internal.execution.impl.steps.CacheStep.execute(CacheStep.java:60)
        at org.gradle.internal.execution.impl.steps.CacheStep.execute(CacheStep.java:37)
        at org.gradle.internal.execution.impl.steps.PrepareCachingStep.execute(PrepareCachingStep.java:33)
        at org.gradle.internal.execution.impl.steps.StoreSnapshotsStep.execute(StoreSnapshotsStep.java:38)
        at org.gradle.internal.execution.impl.steps.StoreSnapshotsStep.execute(StoreSnapshotsStep.java:23)
        at org.gradle.internal.execution.impl.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:95)
        at org.gradle.internal.execution.impl.steps.SkipUpToDateStep.lambda$execute$0(SkipUpToDateStep.java:88)
        at java.util.Optional.map(Optional.java:215)
        at org.gradle.internal.execution.impl.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:52)
        at org.gradle.internal.execution.impl.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:36)
        at org.gradle.internal.execution.impl.DefaultWorkExecutor.execute(DefaultWorkExecutor.java:34)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:109)
        at org.gradle.api.internal.tasks.execution.ResolveIncrementalChangesTaskExecuter.execute(ResolveIncrementalChangesTaskExecuter.java:84)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:92)
        at org.gradle.api.internal.tasks.execution.ResolveBeforeExecutionStateTaskExecuter.execute(ResolveBeforeExecutionStateTaskExecuter.java:75)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:110)
        at org.gradle.api.internal.tasks.execution.ResolveBeforeExecutionOutputsTaskExecuter.execute(ResolveBeforeExecutionOutputsTaskExecuter.java:67)
        at org.gradle.api.internal.tasks.execution.ResolveAfterPreviousExecutionStateTaskExecuter.execute(ResolveAfterPreviousExecutionStateTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:94)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:95)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:63)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:49)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:46)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:46)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:43)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:355)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:343)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:336)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:322)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$1.execute(DefaultPlanExecutor.java:134)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker$1.execute(DefaultPlanExecutor.java:129)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:202)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:193)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:129)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
        at java.lang.Thread.run(Thread.java:748)
org.jetbrains.dokka.javadoc.ClassDocumentationNodeAdapter cannot be cast to com.sun.javadoc.AnnotationTypeDoc

BUILD SUCCESSFUL in 1s
```

Using versions:

```
------------------------------------------------------------
Gradle 5.3
------------------------------------------------------------

Build time:   2019-03-20 11:03:29 UTC
Revision:     f5c64796748a98efdbf6f99f44b6afe08492c2a0

Kotlin:       1.3.21
Groovy:       2.5.4
Ant:          Apache Ant(TM) version 1.9.13 compiled on July 10 2018
JVM:          1.8.0_202 (Oracle Corporation 25.202-b08)
OS:           Mac OS X 10.14.3 x86_64
```

Note that creating html or markdown format docs works correctly. Note also that unlike several other logged issues this is about a _declared_ annotation, not just a _referenced_ one.

**Comment by KrystianUjma** [comment]
fixed, will create PR.

**Comment by roxspring** [comment]
Fixed? Closed?? where's the pull request? which release can I test the fix in?

**Comment by KrystianUjma** [comment]
You can test and use our [multiplatform branch](https://github.com/Kotlin/dokka/tree/multiplatform-support), here is [commit](https://github.com/Kotlin/dokka/commit/c605d0c2b75cce9ad8b818cfd75b49c684e2eb98) which fixes that issue.
We will release it very soon in 0.9.19.

**Comment by jhansche** [comment]
@KrystianUjma any word on 0.9.19?

**Comment by KrystianUjma** [comment]
We're in a testing phase, takes a bit longer than expected.

**Comment by jhansche** [comment]
Got it, thanks for the information.

Would it be reasonable to make a 0.9.19 that has some of these high priority blockers in it (like: you can't generate javadoc if your project contains an annotation, seems to qualify), so that the multiplatform rewrite could become something like 0.10.0, and the internal testing of multiplatform can continue on your schedule?

We're kind of stuck not being able to use _anything_ right now. The released version has a major bug, and the last response was 2 months ago saying it would be released "very soon".  There's no migration guide yet from what I can see, to migrate to the multiplatform snapshot, so it isn't straight forward to just clone and build the multiplatform-support branch, because there are breaking API changes that haven't been documented yet.  So that leaves the only other option being:  clone the master branch, and cherry-pick the linked `commit` from above -- that does work, somewhat, but even then there are still some ClassCastException errors if your source has a nested annotation class (fixed if you move the annotation to an upper level). I was hoping the above linked commit would fix that, but it didn't. And I don't know if it's fixed somewhere else because the multiplatform branch is _hundreds_ of commits beyond master.

The error messages aren't clear when the exception happens, so I have choice other than just looking for "nested annotations" because that seems to be the biggest offender right now.  And then I fix that, and try again, and move onto the next.  It's very slow going.

There also aren't many alternatives to Dokka --  even the Gradle Javadoc task produces no output.

I'm not trying to be critical, I just want to explain why this is a big deal for some of us :(

**Comment by kamilok1965** [comment]
> Would it be reasonable to make a 0.9.19 that has some of these high priority blockers in it 

I'd say not at this point. I think it would take more time than releasing the 0.9.19 with multiplatform support, because we would need to create another version, cherry pick some changes, and test it anyway

> you can't generate javadoc if your project contains an annotation

It was fixed in 0.9.19 some time ago, @KrystianUjma linked to a commit 

> there are still some ClassCastException errors if your source has a nested annotation class

That's surprising, we'll look into that, thanks

> There's no migration guide yet from what I can see, to migrate to the multiplatform snapshot, so it isn't straight forward to just clone and build the multiplatform-support branch, because there are breaking API changes that haven't been documented yet.

That's true, however the README on the `multiplatform-support` branch is pretty up-to-date, so you can see the changes there. Basically some config params were moved to a `configuration` block, so migrating shouldn't be that hard without the migration guide, which we will provide shortly 

> And I don't know if it's fixed somewhere else 

Not to my knowledge

> I'm not trying to be critical, I just want to explain why this is a big deal for some of us :(

No problem, feedback is always appreciated :)

Maybe I can explain the current situation a little. The reason some of major features and bugs are not implemented/fixed is that we're refactoring Dokka for easier extensibility in 0.9.20, which is an important step. After that, many features should be easier and quicker to implement and bugs easier to fix. Also we don't want to repeat the same work twice, which is why some "blockers", as you call them, will be fixed in the 0.9.20 

We are sorry for the inconvenience caused

**Comment by jhansche** [comment]
> That's surprising, we'll look into that, thanks

Thanks - I've seen at least 3 separate issues (this, #477, #455) related to annotations and ClassCastException, but my particular issue I've been chasing down (whack-a-mole style) doesn't seem to be any of those, but looks closely related to this: https://stackoverflow.com/a/56670154/231078
That self-answer says he fixed it by moving the annotations to the top level, to a new package, and suppressing that package.  I _did not_ have to suppress the annotations in any way, but moving them from nested to top-level _does_ fix the errors...  I just have to run the job, wait for failure, grep for what looks like an annotation, and then move it.

If you'd like, I can open a new issue for the one I'm experiencing?  It's worth noting that unlike the other issues linked above, when I encounter it, I don't get a full stack trace -- the trace just ends with `Caused by: ClassCastException` and **nothing else** after that (not even explaining class A cannot be cast to class B, etc).  `--full-stacktrace` also did not help with that.

> That's true, however the README on the multiplatform-support branch is pretty up-to-date, so you can see the changes there. Basically some config params were moved to a configuration block, so migrating shouldn't be that hard without the migration guide, which we will provide shortly

I saw the README, but it wasn't enough to fix my issues.  I'm using the `dokka-android` plugin, and when I tried to set for example `dokka { jdkVersion = 8 }`, I got errors in Gradle configuration, because it says there is no jdkVersion property on the DokkaAndroidTask class.  I eventually gave up on using the multiplatform snapshot I built, because I wasn't able to fix the configuration issues.

Thanks for the information and for explaining the project's plan.
